### ISSUE TITLE
Remove all global css pollution

### ISSUE THREAD

**Issue by Airkro** [facebook/docusaurus#6032]
### Have you read the Contributing Guidelines on issues?

- [X] I have read the [Contributing Guidelines on issues](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#reporting-new-issues).

### Motivation

Global css messes up many things, please consider removing them (rewrite in classes), it will save the day.

![image](https://user-images.githubusercontent.com/12206891/144028773-ab9750c4-1a43-450d-a605-a3e32f235e07.png)

### Self-service

- [ ] I'd be willing to do some initial work on this proposal myself.

**Comment by Josh-Cena** [comment]
Agree that this is a valid concern that I have been asked on Discord. However, several things missing from the proposal that I wish we can get response on:

1. **Impact.** Is there a really strong use-case that's hindered by the global CSS? Won't `all: revert` work if you want a part of the page unstyled?
2. **Possible implementation.** How would we work to implement your proposal? Many of these elements are directly generated from Markdown (e.g. `<kbd>`), which is also the primary focus of Docusaurus. Does that mean we should map all Markdown elements to one with a class name?

> Global css messes up many things

This is an assertion and I don't see exact what it messes up

> please consider removing them (rewrite in classes)

You should report that on the [Infima](https://github.com/facebookincubator/infima) side, the CSS framework developed alongside Docusaurus. In Docusaurus we are more concerned about how CSS classes can interoperate with the current way that HTML is generated.

**Comment by slorber** [comment]
IMHO we could scope those rules under a `.infima-app` class or something used at the root of the Docusaurus app.

But would this solve your problem @Airkro ?

**Comment by Josh-Cena** [comment]
> we could scope those rules under a `.infima-app` class or something used at the root of the Docusaurus app.

The OP wants to easily opt-out from Infima. Scoping probably doesn't prevent the tag selectors from being applied to every element within the `<Layout>`

**Comment by tobemaster56** [comment]
I meet this problem too

We are building our own React component library. There are a lot of documents based on docusaurus. Thanks to the mdx of docusaurus, we can easily render React components in the document. We need to create a lot of demos. We have a [pagination component]( http ://design.shuyun.com/portal/docs/components/navigation/pagination), the `li` element is used, but in the demo, the style is a bit abnormal. I found out that it was a problem with the global styles of docusaurus.

I created a discussion [css style isolation](https://github.com/facebook/docusaurus/discussions/6396) about it,  with detailed instructions and screenshots, and I tried `all:revert`, but it didn't work.

While we can override the relevant styles to solve this kind of problem temporarily, I really feel that this is not the best solution and is a bit ugly. Soon after I discovered the problem, another colleague had the same problem, a calendar demo in which the style of the tables in the component was contaminated, and it took him a lot of time to find out the cause.

I hope there is a once and for all way to be free from global styles.

After discussing it with @Josh-Cena , I tried to use [react-shadow-root]https://www.npmjs.com/package/react, but in the end, it failed, and for our component library, our styles were also global and blocked by shadow-dom. I gave up

In the end, I still hope that the official will come up with an elegant plan, and I hope to hear a response as soon as possible.

**Comment by slorber** [comment]
Agree, it should be possible to mount a component library in Docusaurus and expect it to not be affected by Docusaurus styles (without requiring the usage of an iframe)

We'll look into this.

In the meantime please report any Docusaurus style that is affecting the styling of your components.

**Comment by Josh-Cena** [comment]
What about having a special class in Infima that reverts all Infima styles? I still find value in certain global styles instead of moving everything to class names.

**Comment by Airkro** [comment]
Sorry for not responding. I integrated Swagger UI with Docusaurus. I have to put Swagger UI in shadow dom to avoid CSS pollution.

**Comment by Josh-Cena** [comment]
Closing this. We can discuss in Infima instead, see also https://github.com/facebookincubator/infima/issues/8

I think the easiest solution is to provide an OOTB `Sandbox` component which is basically a shadow DOM? A `reset` utility class can do the job as well.

**Comment by slorber** [comment]
I'll re-open because we can also do something on Docusaurus side, such as providing another theme (likely Tailwind) that does not have this problem in the first place.

It should be possible to mount MDX components in your doc and have those components not affected by Docusaurus theme unscoped styles.

Now we can also open another discussion on Infima to ensure all styles there are scoped properly.

**Comment by Josh-Cena** [comment]
I still don't understand this. First is "why", second is "how".

Infima is more comparable to Bootstrap than Tailwind. We don't ask users to compose class names to get any useful styles; we directly target elements. If user write `<kbd>foo</kbd>` she should immediately get styling that looks consistent with the rest of the design, instead of having to compose layers of class names to adjust the margin, the border, the color... Its goal of driving _content-centric_ website effectively _means_ users should stop caring about styling in all and delegate everything to the default styles.

Let's say the use-case of showcasing custom components is justified. How do we scope Infima? We inevitably have to do it "the Tailwind way", i.e. use class names instead of tag selectors everywhere. But does that sound like the best solution?

It instantly means we have to do A LOT more MDX component mapping because literally everything, from `ul` to `table` to `hr`, has Infima-default styles.

Furthermore, it means React pages also have to have extra class names just to get the same styling as Markdown pages. I already have a hard time explaining to people that MDX != HTML != JSX because there can be build-time transformations. Simply transforming asset links has already confused users when `[link](./asset.pdf)` works in Markdown but `<a href="./asset.pdf">link</a>` doesn't work in JSX. (Which led to my doc addition that [Markdown is declarative](https://docusaurus.io/docs/next/markdown-features#:~:text=Markdown%20is%20declarative)). Now we not only want to take the intuition of Markdown -> JSX mapping away, but also the _JSX -> JSX_ mapping intuition away. Being intuitive—not being perfect for every use-case—is what makes us competitive.

Now, can this be done WITHOUT "remove all global css polution"? Yes. We can do it in userland today with iframes and/or shadow DOM and/or a `reset` class. Those all sound like viable use-cases.

Taking a step further back, how can Docusaurus help with this use-case? This should be discussed in Infima instead of Docusaurus. Until we decide which path we want to take within an Infima thread, there doesn't seem to be anything Docusaurus can do, if we need to do anything at all.

**Comment by slorber** [comment]
> It instantly means we have to do A LOT more MDX component mapping because literally everything, from ul to table to hr, has Infima-default styles today. We have to transform them to something like <table className="table" />.

We'll likely have to do all that anyway if we want to support a tailwind theme, unless you also want Tailwind to apply global classes?

> We can do it in userland today with iframes and/or shadow DOM and/or a reset class. 

- iframes is not really a good solution. People want to embed native runnable code in their doc.
- shadow DOM: looks like an interesting
- reset: why not but then do we want users to have to maintain this class themselves over time?


---

I'm open to exploring the shadow DOM solution.

Maybe the issue is not global CSS, but simply the inability to embed natively your lib's code in your doc without having any styling conflicts?

If we have a solution to this problem, with documentation and some dogfooding, maybe we don't have to handle the global unscoped CSS issue

**Comment by Josh-Cena** [comment]
Yeah, if we decide to go with shadow DOM it could be most relevant to Docusaurus (we can even offer a theme component like `Sandbox`). I think someone did tell me they solved it this way.

> unless you also want Tailwind to apply global classes?

That I'm actually not sure. Tailwind is really reusable styles rather than reusable components (which Bootstrap/MUI/Infima are more of). To some degree, that contradicts with the goal of a content-centric rather than design-centric site. Maybe I'm okay with the Tailwind theme doing more MDX component mapping? Certainly don't want it to be a widespread pattern though.

**Comment by Josh-Cena** [comment]
Also note that if a user opts into Tailwind I'd assume she's already aware of the Tailwind system (which may not be at all apparent to other casual developers), so the JSX/MDX behavior deviation may be less important. But for the classic theme, which I still aim to be the most accessible for everyone in the long term, should be intuitive enough.

**Comment by gpbl** [comment]
Thanks for reopening this! 

Please note this is not just related to the markdown rendering. Even _outside_ the markdown docs, there are selectors polluting the global scope, which affects also regular pages/routes. So I'm not sure if the "shadow" solution would work.

For example, this is a route in Docusaurus: https://react-day-picker.js.org/render?example=start. The style is not correct because the CSS overrides the table head style:

```css
table th {
  background-color: var(--ifm-table-head-background);
  color: var(--ifm-table-head-color);
  font-weight: var(--ifm-table-head-font-weight);
}
```

We shouldn't see these selectors in the CSS. However, reverting them would affect the existing implementations, so fixing this is impracticable. 

Maybe it's time to develop a new theme? I evaluated https://docusaurus-theme-no-style.netlify.app, to start  implementing my own, but that is a lot of work.

**Comment by gpbl** [comment]
> Now, can this be done WITHOUT "remove all global css polution"? Yes. We can do it in userland today with iframes and/or shadow DOM and/or a reset class. Those all sound like viable solutions that (a) is extensible (b) is robust (c) doesn't require any work from Docusaurus

@Josh-Cena those are not viable solutions. Iframe is still affected by CSS pollution, and a "reset" class will never work well. To me, writing the CSS that way has been a bad choice from start, and it has nothing to do with MDX mapping or the "tailwind" way.

I understand that Infima intends to replicate Bootstrap philosophy (even if Bootstrap splits the global CSS classes in a "reboot" CSS). I question if Infima is the right choice for Docusaurus: importing a components in MDX, and having its style overridden by Docusaurus, makes the MDX feature much less useful.

**Comment by Josh-Cena** [comment]
Shadow DOM should be the way to go, because that's literally the purpose of shadow DOM. I didn't investigate, but I do know of successful integration cases.

**Comment by gpbl** [comment]
> Infima is developed for Docusaurus, so asking if Infima is the right choice is asking if theme-classic is the right choice

Infima has tons of quirks, I've spent countless hours trying to workaround them. It is not the right theme for documenting a components library – in this case, it would be nice a note in the docs. Anyway here we are, reporting our issues with it. Hopefully soon we can have an alternative theme!

**Comment by Josh-Cena** [comment]
Yes, Infima is not designed with component library docs in mind—we do not expect such use-case. Using an alternative theme like Tailwind should be able to fix it.

**Comment by gpbl** [comment]
> Yes, Infima is not designed with component library docs in mind—we do not expect such use-case.

I can spot few use cases in the Showcase gallery - e.g https://ionicframework.com/docs/components 
I'm sure they just run into our issues, give up and/or solved it using Codesandbox. 

The case of react-day-picker.js.org is not even a component library - is just a single component.

**Comment by Josh-Cena** [comment]
Ionic solved it through shadow DOM which is what I've been saying all along:

<img width="251" alt="image" src="https://user-images.githubusercontent.com/55398995/164483889-b619a705-ed09-4978-b55b-0d951c8d5d6a.png">

Plus they are a native component library so they have a different battle.

**Comment by slorber** [comment]
Looks like upcoming CSS `@scope` this could be a good solution:

<img width="629" alt="CleanShot 2022-06-06 at 13 08 31@2x" src="https://user-images.githubusercontent.com/749374/172149707-f46dfb25-1b38-4e78-84f9-b826584eb301.png">

https://twitter.com/sebastienlorber/status/1533767305875824641

**Comment by TimonVS** [comment]
Our team has been using [ReactShadow](https://github.com/Wildhoney/ReactShadow) to workaround this issue. But we haven't been able to get it to work with Emotion yet (tried creating a separate cache and having the styles be inserted in the shadow root, but haven't managed to get it to work yet).

**Comment by stevenpetryk** [comment]
Just reporting here that at Discord we've had some success using shadow DOM as well. We had to tell webpack to split the bundle into three CSS files:

- docusaurus-styles.css (all the styles inside our doc site)
- discordapp-styles.css (all the app styles, except for the reset styles)
- discordapp-reset.css (our own app's reset)

In the `head`, we load `docusaurus-styles.css` and `discordapp-styles.css` (because we want to use our components in doc pages), but in the shadow DOM, we _only_ load `discordapp-*.css` (because we don't want Docusaurus' global styles to pollute the component playgrounds, which need to look _exactly_ the same as they do in Discord).

The only issue we've faced with Shadow DOM is that some event handlers needed to be changed. For example, for a popup component (think popper, context menus, dropdowns, etc.), we had a `mouseup` listener on `document`, and it checked if `popupContainer.contains(event.target)`. With shadow DOM, this doesn't work, because `event.target` is always the shadow root, so we had to change it to `popupContainer.contains(event.composedPaths()[0])`. 

I'll report back if I see other issues. I agree that it would be great to not have to use Shadow DOM, but it sounds _far worse_ to have to import some `<Pre>` component rather than just using HTML `<pre>`. `@scope` is really the answer.

**Comment by slorber** [comment]
A new interesting solution from @bluwy 

[Whyframe](https://whyframe.dev/) permits to easily isolate the rendering of a React component inside an iframe.

There's a [Docusaurus plugin](https://whyframe.dev/docs/integrations/docusaurus) to try

**Comment by dcousineau-godaddy** [comment]
We took an approach similar to @stevenpetryk and wanted to post the rough outline for anyone else who finds this thread below.

On a side note, I think there's a possibility that a contribution to `@docusaurus/theme-live-codeblock` _could_ be made to have a toggleable Shadow DOM approach, the primary coordination _could_ be that Docusaurus's `splitChunks` configuration for CSS files is altered such that `infima` and `@docusaurus` css files are chunked separately, and "everything else" is in a separate well-named chunk that, if shadow dom is enabled in the live codeblock, could be pulled in whole-sale for a ... "quick and dirty" isolated codeblock experience. Will see if I can carve out some time to put together the skeleton of the idea.

---

In the theme that we ship, our basic approach was to tweak the splitChunks cacheGroups to isolate _our_ component library styles:

```javascript
const DEFAULT_LIVE = {
  enableShadowDOM: true,
  includeSplitChunks: c => {
    return /@package-namespace[\\/].+\.s?css/.test(c.identifier());
  }
};

module.exports = function (context, options) {
  const { /* ...snip...*/, live = DEFAULT_LIVE } = options || {};

  return {
    name: 'our-theme',
    // ...snip...
    configureWebpack(config, isServer, utils) {
      const updateConfig = {};

      if (!isServer && live.enableShadowDOM) {
        // Utilize split chunks to ensure CSS assets from Docusaurus are separated from others
        updateConfig.mergeStrategy = updateConfig.mergeStrategy ?? {};
        updateConfig.mergeStrategy['optimization.splitChunks.cacheGroups'] = 'merge';
        updateConfig.optimization = {
          ...updateConfig?.optimization,
          splitChunks: {
            cacheGroups: {
              liveStyles: {
                name: 'styles-live',
                test: c => {
                  // Additionally include any styles that may have been pulled in by ReactLiveScope overrides
                  return /ReactLiveScope[\\/].+\.s?css/.test(c.identifier()) || live.includeSplitChunks(c);
                },
                chunks: 'all',
                // enforce: false feels weird to have, but it DOES result in the CSS outputs we need.
                // @see https://github.com/thecrypticace/webpack-css-chunks-test/blob/master/webpack.config.js
                enforce: false,
                priority: 60 // The default `styles` cacheGroup from Docusaurus is set to priority 50
              }
            }
          }
        };
      }

      return updateConfig;
    },
    // ...snip...
  };
}
```

And then we ejected the `Playground` component from `@docusaurus/theme-live-codeblock` into our custom theme and made some tweaks to mount the `<LivePreview />` in a shadow root (there is a lot of snipping below, we have a heavily customized Playground component)

```jsx
// @see https://github.com/facebook/docusaurus/blob/master/packages/docusaurus-theme-live-codeblock/src/theme/Playground/index.js
import React, { useState, useRef, useEffect, useMemo } from 'react';
import { LiveProvider, LiveEditor, LiveError, LivePreview } from 'react-live';
import useDocusaurusContext from '@docusaurus/useDocusaurusContext';
import ShadowRoot from 'react-shadow';

export default function Playground({ children, transformCode, ...props }) {
  // ...snip...
  const {
    siteConfig: {
      themeConfig: { live = {} }
    }
  } = useDocusaurusContext();

  // We assume that shadowDOM is enabled by default
  const enableShadowDOM = live?.enableShadowDOM ?? true;

  const sheets = useMemo(() => {
    if (!enableShadowDOM) return [];

    const sheets = typeof document !== 'undefined' && document?.styleSheets?.length && Array.from(document.styleSheets);
    if (!sheets) return [];

    return sheets.filter(sheet => {
      if (!sheet.href) return false;
      const sheetHref = new URL(sheet.href);
      return sheetHref.pathname.match(/styles-live/i);
    });
  });

  return (
    <div className={styles.playgroundContainer}>
      <LiveProvider /* ...snip... */ >
        <div className={styles.playgroundPreview}>
          {enableShadowDOM ? (
            <ShadowRoot.div ref={root} mode='open'>
              <LivePreview />
              {sheets.map(sheet => (
                <style type='text/css'>{Array.from(sheet.cssRules, rule => rule.cssText || '')}</style>
              ))}
            </ShadowRoot.div>
          ) : (
            <LivePreview />
          )}
          {showEditor && <LiveError />}
        </div>
        <LiveEditor className={styles.playgroundEditor} />
        {/* ...snip... */}
      </LiveProvider>
    </div>
  );
}

```

> **Note**
> The above approach is currently naïve especially in regards to pulling in the stylesheet for the live playground (effectively only once on mount and maybe a second time if React clears the memoization cache).
> We'll be spending more effort over time improving / bullet proofing the solution but for now it works "good enough".

**Comment by dcousineau** [comment]
We took an approach similar to @stevenpetryk and wanted to post the rough outline for anyone else who finds this thread below.

On a side note, I think there's a possibility that a contribution to `@docusaurus/theme-live-codeblock` _could_ be made to have a toggleable Shadow DOM approach, the primary coordination _could_ be that Docusaurus's `splitChunks` configuration for CSS files is altered such that `infima` and `@docusaurus` css files are chunked separately, and "everything else" is in a separate well-named chunk that, if shadow dom is enabled in the live codeblock, could be pulled in whole-sale for a ... "quick and dirty" isolated codeblock experience. Will see if I can carve out some time to put together the skeleton of the idea.

---

In the theme that we ship, our basic approach was to tweak the splitChunks cacheGroups to isolate _our_ component library styles:

```javascript
const DEFAULT_LIVE = {
  enableShadowDOM: true,
  includeSplitChunks: c => {
    return /@package-namespace[\\/].+\.s?css/.test(c.identifier());
  }
};

module.exports = function (context, options) {
  const { /* ...snip...*/, live = DEFAULT_LIVE } = options || {};

  return {
    name: 'our-theme',
    // ...snip...
    configureWebpack(config, isServer, utils) {
      const updateConfig = {};

      if (!isServer && live.enableShadowDOM) {
        // Utilize split chunks to ensure CSS assets from Docusaurus are separated from others
        updateConfig.mergeStrategy = updateConfig.mergeStrategy ?? {};
        updateConfig.mergeStrategy['optimization.splitChunks.cacheGroups'] = 'merge';
        updateConfig.optimization = {
          ...updateConfig?.optimization,
          splitChunks: {
            cacheGroups: {
              liveStyles: {
                name: 'styles-live',
                test: c => {
                  // Additionally include any styles that may have been pulled in by ReactLiveScope overrides
                  return /ReactLiveScope[\\/].+\.s?css/.test(c.identifier()) || live.includeSplitChunks(c);
                },
                chunks: 'all',
                // enforce: false feels weird to have, but it DOES result in the CSS outputs we need.
                // @see https://github.com/thecrypticace/webpack-css-chunks-test/blob/master/webpack.config.js
                enforce: false,
                priority: 60 // The default `styles` cacheGroup from Docusaurus is set to priority 50
              }
            }
          }
        };
      }

      return updateConfig;
    },
    // ...snip...
  };
}
```

And then we ejected the `Playground` component from `@docusaurus/theme-live-codeblock` into our custom theme and made some tweaks to mount the `<LivePreview />` in a shadow root (there is a lot of snipping below, we have a heavily customized Playground component)

```jsx
// @see https://github.com/facebook/docusaurus/blob/master/packages/docusaurus-theme-live-codeblock/src/theme/Playground/index.js
import React, { useState, useRef, useEffect, useMemo } from 'react';
import { LiveProvider, LiveEditor, LiveError, LivePreview } from 'react-live';
import useDocusaurusContext from '@docusaurus/useDocusaurusContext';
import ShadowRoot from 'react-shadow';

export default function Playground({ children, transformCode, ...props }) {
  // ...snip...
  const {
    siteConfig: {
      themeConfig: { live = {} }
    }
  } = useDocusaurusContext();

  // We assume that shadowDOM is enabled by default
  const enableShadowDOM = live?.enableShadowDOM ?? true;

  const sheets = useMemo(() => {
    if (!enableShadowDOM) return [];

    const sheets = typeof document !== 'undefined' && document?.styleSheets?.length && Array.from(document.styleSheets);
    if (!sheets) return [];

    return sheets.filter(sheet => {
      if (!sheet.href) return false;
      const sheetHref = new URL(sheet.href);
      return sheetHref.pathname.match(/styles-live/i);
    });
  });

  return (
    <div className={styles.playgroundContainer}>
      <LiveProvider /* ...snip... */ >
        <div className={styles.playgroundPreview}>
          {enableShadowDOM ? (
            <ShadowRoot.div ref={root} mode='open'>
              <LivePreview />
              {sheets.map(sheet => (
                <style type='text/css'>{Array.from(sheet.cssRules, rule => rule.cssText || '')}</style>
              ))}
            </ShadowRoot.div>
          ) : (
            <LivePreview />
          )}
          {showEditor && <LiveError />}
        </div>
        <LiveEditor className={styles.playgroundEditor} />
        {/* ...snip... */}
      </LiveProvider>
    </div>
  );
}

```

> **Note**
> The above approach is currently naïve especially in regards to pulling in the stylesheet for the live playground (effectively only once on mount and maybe a second time if React clears the memoization cache).
> We'll be spending more effort over time improving / bullet proofing the solution but for now it works "good enough".

**Comment by gpbl** [comment]
Thanks for the feedbacks! I am still not happy with the workarounds described here.

What we need is the ability to create a blank HTML page in Docusaurus, but _without any styles_. We could them link this page with a standard `iframe`. Can we do this without forking the whole classic theme?

**Comment by slorber** [comment]
> What we need is the ability to create a blank HTML page in Docusaurus, but without any styles. We could them link this page with a standard iframe. Can we do this without forking the whole classic theme?

If you don't need SPA navigation, you can add any html file in `/static/xyz.html` and link to it with `pathname:///xyz.html` or display it inside an iframe

**Comment by slorber** [comment]
Note to myself: cascade layers can also be a good alternative to opt-out of the style we apply by default.

This article explains how to use `all: revert-layer` for that purpose:
https://www.mayank.co/blog/revert-layer

**Comment by bspeice** [comment]
One example where I think the global CSS introduces some display problems is with `theme-live-codeblock`. The global `pre { border-radius: var(--ifm-pre-border-radius); }` produces a visual gap where the "LIVE EDITOR" title bar and code block meet:

![image](https://github.com/user-attachments/assets/b34bdd01-8437-462e-8afc-c3a2ac9d207c)

Zooming in a bit:

![image](https://github.com/user-attachments/assets/e14fc3fe-fdcf-4fe4-9c59-c990dc68c2ef)

I've ejected the component so I can modify `styles.module.css` locally and disable the `border-radius`.

**Comment by michaelwarren1106** [comment]
+1000 for CSS cascade layers. If all the Infima styles are wrapped in a layer regardless of how they are written, it would be MUCH more straightforward to override them without adding more css or trying revert.

I'm experiencing this right now. I have a CSS library package with only layered CSS in it. All of the docusaurus styles from Infima are unlayered. In browsers, styles NOT in a cascade layer will ALWAYS be higher in the specificity stack than layered styles. So i cannot get my layered styles to apply over top the unlayered styles from Infima. There's no path to success without using patch-package and manually re-rewriting the Infima in my app.

If Infima was just layered and that layer name was documented, then any style from Infima could be overridden simply by making a new cascade layer and ordering it on top of whatever the Infima/Docusaurus style layer(s) is/are. Then there's no problem, no need for revert, no need for Infima or docusaurus to change the way styles are written, and global styles from Infima still automatically apply if there's no conflict in a higher layer just like it was mentioned at the start of this issue was an important concern.

That said, is there any way that I can wrap Infima styles in a layer through Docusaurus features? I could test it and report back if so.

**Comment by michaelwarren1106** [comment]
For some reference, Starlight, a docusaurus alternative for Astro uses Cascade layers for this exact purpose:

https://starlight.astro.build/guides/css-and-tailwind/#cascade-layers

**Comment by slorber** [comment]
I'll try to see soon if we can provide an opt-in flag to add a cascade layer.

If this produces no visual change to our website visual regression tests, and no bug in community sites, maybe we could turn this on automatically for v4.

**Comment by stevenpetryk** [comment]
This issue isn't about precedence, it is Docusaurus' element selectors and inherited styles.

**Comment by michaelwarren1106** [comment]
it’s both. if docusaurus wraps all of its styles in a layer then it doesnt matter whether the styles are element selectors or not. just being layered at all means that whatever the selector is and however specific it is, it can always be overridden by either unlayered styles or by styles in a higher layer. 

```


// docu styles
@layer docu {
   h1 { color:red !important;
}

//app custom styles
@layer docu, app;


@layer app {
  h1 { color: yellow !important; // wins over docu
}


h1 { color: blue;} //wins over docu and app
.something h1 { color:green } // also wins over docu and app
```

**Comment by slorber** [comment]
I created a POC PR here: https://github.com/facebook/docusaurus/pull/11142

I found a way to conditionally apply a CSS layer based on a config flag and specific file CSS paths. Applying the layer doesn't seem to break things, so I guess it could be fine to ship in v4, and through a new v4 future flag to release in v3.8.

I guess we could use the following layers by default, does it make sense?

```css
@layer infima, theme-common, theme-classic;
```

Considering, for now the layers are only opt-in, it's probably unnecessary to create a layer for the user app and its custom CSS. Later, we could eventually add an `@app` layer and add it by default to our new init template. Let me know what you think?

Note: as you can see in the PR, this can be implemented as a simple postCSS plugin, and even as a Docusaurus plugin using the `configurePostCss()` lifecycle hook. So technically, you can already add this to your own website through our plugin APIs today.

--- 

@stevenpetryk can you give a concrete example where layers wouldn't let you override the default CSS?

---

Layers aren't going to solve global CSS pollution alone, but should let you do so on your own.

I'll need to experiment a bit to see it in practice, but `all: revert-layer` might help you do so, according to these resources:

- https://mayank.co/blog/css-reset-layer/
- https://mayank.co/blog/revert-layer/

---

Can you please submit cases where that global CSS messed up with what you tried to do? 

If you can show me small standalone components that did not render as you wanted, I could try to create test cases that we'll use to ensure that we are now able to render these components as you expect, and capture that in our visual regression test suite.

**Comment by michaelwarren1106** [comment]
the poc looks good!

the one comment i have is whether or not the base layer and the theme layers need to be ordered in the docusaurus styles or if they can be left unordered? i’m not 100% clear on how ordered layers can be re-ordered. i assume it could be done with another `@layer` order declaration later in the style bundle, but not sure. 

if docusaurus can function by establishing layers but without ordering layers by default i think that would be best? then maybe a little bit of docs about the recommended layer order?

**Comment by slorber** [comment]
> the one comment i have is whether or not the base layer and the theme layers need to be ordered in the docusaurus styles or if they can be left unordered?

I'd recommend reading this great article, it's worth it and covers various interesting things: https://css-tricks.com/css-cascade-layers/

Cascade layers stack in the order they first appear.

This is why the order is often defined at the very top, without any CSS rule:

```css
@layer reset, defaults, components, utility
```

To allow users to provide their own layer order (and eventually interleave their own layers with ours), we need to enable them to provide such a top declaration. 

`custom.css` won't do the trick because those rules will appear later in the global stylesheet. This behavior has been implemented on purpose so that custom styles take precedence over our styles to allow overriding them, so we can't change this easily.


> if docusaurus can function by establishing layers but without ordering layers by default i think that would be best? then maybe a little bit of docs about the recommended layer order?

We can establish layers, but we absolutely want to give an explicit order. It makes no sense if Infima classes would override CSS modules from our theme for example. 

We always want Infima to appear first so that users can easily override it by default, even if they use layers in their own code.

**Comment by slorber** [comment]
## Cascade Layers POC and results

I've been using @gpbl library [React Day Picker](https://daypicker.dev/) (PR https://github.com/gpbl/react-day-picker/pull/2764) as a way to test CSS isolation using CSS cascade layers. 

The result seems pretty good, and I can revert Infima to create the demo. 

However the site also provided its own global pollution within `custom.css` with selectors such as `.markdown table {}`. In such case, to isolate your demo, you also need to revert those styles, and wrapping them in a dedicated layer such as `website` is helpful so that you can revert your own pollution too.

To revert, you can use a class like this one:

```css
.my-demo:not(#a#b) {
  &,
  * {
    @layer infima {
      all: revert-layer;
    }

    @layer website {
      all: revert-layer;
    }
  }
}
```

Then you can apply the `.my-demo` class to any container to get rid of Infima and your own global styles.

The `:not(#a#b)` selector is important here (called the ["impossible id selector"](https://mayank.co/blog/revert-layer/), because we want the rule `all: revert-layer` to have high "id" specificity to take precedence over all the other selectors in the layer, otherwise some rules won't be reverted. 

Each layer needs to be reverted one by one (and using `revert-layer !important` won't do that FYI, [see why](https://css-tricks.com/css-cascade-layers/#aa-reverting-important-layers))

---

There are also considerations regarding the usage of `!important` in layers: if Infima is the first layer to appear, it's the least powerful one (easy to override), but using `!important` in Infima makes rules being more powerful than `!important` in other layers and unlayered rules. We have a few `!important` in Infima, so if we layer it, those rules will now become harder to override (need to create a layer before Infima and use `!important` in it). 

I don't think this will be a problem in practice, considering many of these rules are utility ones, and should "always win" (similar to Tailwind, style driven by the HTML markup). Examples include `.text--break`, `.shadow-<level>` and `margin-<side>-<size>`. Apart from these, we have only 3 cases of non-utility `!important` usage that could become harder to override, and can work on removing them.

---

## Cascade Layers in Docusaurus

For the initial Docusaurus integration, I'd like to keep things simple and only create a layer for Infima. 

Our theme classic uses CSS modules, so apart from a few exceptions, most styles are already scoped so it doesn't seem super useful. Also, we should take into consideration that the layer is applied dynamically (opt-in) with postcss, and the swizzled components in `website/src/theme` should be taken into consideration (for both newly swizzled components, and existing ones). 

Also, we have many other packages providing theme components, and it's not clear how to layer them and define an explicit order automatically. It's even more complex to do this globally for third-party packages.

It's also not possible to automatically wrap the site's custom CSS (usually `custom.css`) in a layer without producing unwanted side effects. This is particularly the case if your custom CSS overrides styles from an external CSS stylesheet that is unlayered (we have the case with Algolia DocSearch, but it might happen with any other third-party package providing a CSS file).


So, the current plan is to:
- Implement an opt-in `infima` layer in Docusaurus v3.8 as an experimental flag
- Provide a built-in `infima-revert-layer` class for demo/playground isolation
- Users should take care of their own CSS pollution. For that, they can create their own `website` layer and revert it too.
- Depending on feedback, we'll decide what we'll do for v3.9 (flag on by default?) and 4.0 (remove flag, always use cascade layers?)

**Comment by slorber** [comment]
Docusaurus v3.8 is out and lets you opt out of Infima / Theme styles: 
https://docusaurus.io/blog/releases/3.8

Please let us know if this solution is good to solve this problem, or if it didn't work for you.

**Comment by michaelwarren1106** [comment]
@slorber i just pulled in 3.8 to give it a shot!

One thing I noticed immediately after installing and trying to add the config for the cascade layers plugin is that both the config options in `v4` are set as required in TS types. Should they be optional since they are flags?

```
// line 134 of config.d.ts
export type FutureV4Config = {
  removeLegacyPostBuildHeadAttribute: boolean;
  useCssCascadeLayers: boolean;
};
```

I only wanted the cascade layer flag, cuz I haven't read up on the `removeLegacyPostBuildHeadAttribute` option, so I'm not sure I wanted to set it. I set it to false for right now, but Id rather just not set it at all.

Just wanted to post this comment quickly, I'll get back with cascade layer testing shortly.

Thanks for this update!
